backport-2.0: cli: Deprecate --wait=live in `node decommission`

tbg opened the pull request:
This is a bit of an unusual backport. The previous behavior is correctly documented, but it simply isn't that useful to operators who want to feel reasonably certain that cluster health is back to green before carrying on with their day. I'm not sure what our policy here should be this late into 2.0, but I'm tempted to say that we're not going to cherry-pick this. @bdarnell, what's your take here?

Backport 1/1 commits from #27027.

/cc @cockroachdb/release

---

Refer to issue #26880.

When you try to decommission a node that is down, today one has to use
`decommission --wait=live`, which does not verify whether the down node
is part of any ranges, and manually verify that the cluster has
up-replicated all ranges elsewhere. This is far from ideal, in particular
since there is no automated way to reliably check this.

`--wait=live` is deprecated and its behaviour replaced by that of
`--wait=all`. Instead of relying on metrics for replica counts, which
may be stale, look at authoritive source - range metadata.

Release note (cli change):
Deprecate `--wait=live` parameter for `node decommission`. `--wait=all`
is the default behaviour. This ensures CockroachDB checks ranges are on
the node to be decommissioned are not under-replicated before the node
is decommissioned.

neeral committed:
cli: Deprecate --wait=live in `node decommission`

Refer to issue #26880.

When you try to decommission a node that is down, today one has to use
`decommission --wait=live`, which does not verify whether the down node
is part of any ranges, and manually verify that the cluster has
up-replicated all ranges elsewhere. This is far from ideal, in particular
since there is no automated way to reliably check this.

`--wait=live` is deprecated and its behaviour replaced by that of
`--wait=all`. Instead of relying on metrics for replica counts, which
may be stale, look at authoritive source - range metadata.

Release note (cli change):
Deprecate `--wait=live` parameter for `node decommission`. `--wait=all`
is the default behaviour. This ensures CockroachDB checks ranges on the
node to be decommissioned are not under-replicated before the node is
decommissioned.

bdarnell commented on pkg/cli/cliflags/flags.go:
```diff
@@ -653,9 +653,6 @@ Takes any of the following values:
   - all:  waits until all target nodes' replica counts have dropped to zero.
     This is the default. Use this unless you are targeting down nodes. In the presence
     of down nodes, this will likely wait forever.
```
Remove the last sentence of the `all` docs.

bdarnell commented on pkg/server/admin.go:
```diff
@@ -1291,21 +1290,30 @@ func (s *adminServer) DecommissionStatus(
+		for _, nodeID := range nodeIDs {
+			replicaCounts[nodeID] = 0
+		}
+		return txn.Iterate(ctx, keys.Meta2Prefix, keys.MetaMax, pageSize,
```
Scanning meta2 is not sufficient - you need to scan meta1 as well to ensure that all meta2 replicas have been moved.

bdarnell reviewed (commented):
I think this is a good change and worth back-porting (with one significant change noted below), although I think it's mis-described in its commit message and release note: the primary effect of this change is not the `cli` change to make `--wait=live` work like `--wait=all`, it's the server-side change to make `--wait=all` useful in the presence of dead nodes. 

In fact, maybe the change should be split: backport the `server` changes to `release-2.0` but leave the `cli` changes for 2.1. CLI changes in patch releases are troublesome because people often use the CLI from a control machine whose cockroach version is not upgraded and monitored as strictly as the servers.

tbg committed:
server: check all meta records during decommissioning

By scanning only the meta2 information, replicas of the meta2 ranges
itself would be ignored. Include meta1 in the scan to address this.

This wasn't released, so no release note has been added.

Release note: None

tschottdorf commented:
@bdarnell PTAL at the last commit - is this what you had in mind? It keeps cli output the same but updates everything else. I reverted to the previous acceptance test which *should* pass.

bdarnell reviewed (approved):
> Reviewed 3 of 8 files at r1, 2 of 3 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27158)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/node.go, line 387 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LH4aY5W3hnXozhcuI8l:-LH4aY5W3hnXozhcuI8m:b-jxmdtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9abd9980349cd9d9b878ac0fe49a755dc38a9bf/pkg/cli/node.go#L387)):*
> ```Go
> 		allDecommissioning := true
> 		for _, status := range resp.Status {
> 			if wait != nodeDecommissionWaitLive || status.IsLive {
> ```

I was thinking that the `cli` package would remain unchanged, so `--wait=live` would have the same (not very useful) semantics as before. 
___
*[pkg/cli/cliflags/flags.go, line 655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LG_TZt4RfWfU8bh_5KS-r1-655:-LH4aQbY9v25hlfKwrq7:b9zjdge) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfae0c2bba2248db8b2090da40b6f99409082264/pkg/cli/cliflags/flags.go#L655)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove the last sentence of the `all` docs.
</blockquote></details>

Keep "This is the default".


<!-- Sent from Reviewable.io -->

tbg committed:
cli: massage the behavior for release-2.0

We want to wait for the (now consistent) replica count
even if --wait=live is specified, though we do not
want to change anything about the output.

Release note: None

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27158)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/cliflags/flags.go, line 655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LG_TZt4RfWfU8bh_5KS-r1-655:-LH4ciyp4NZXc5DhVQI9:b7rz6pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfae0c2bba2248db8b2090da40b6f99409082264/pkg/cli/cliflags/flags.go#L655)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Keep "This is the default".
</blockquote></details>

Done. You didn't want to keep the original text here, right?
___
*[pkg/server/admin.go, line 1300 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LG_TZt4RfWfU8bh_5KW-r1-1300:-LH4eZbU8RRlmWvlGuQK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfae0c2bba2248db8b2090da40b6f99409082264/pkg/server/admin.go#L1300)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Scanning meta2 is not sufficient - you need to scan meta1 as well to ensure that all meta2 replicas have been moved. 
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 387 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LH4aY5W3hnXozhcuI8l:-LH4cjQqA-QTCfOcFBNI:buzel3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9abd9980349cd9d9b878ac0fe49a755dc38a9bf/pkg/cli/node.go#L387)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I was thinking that the `cli` package would remain unchanged, so `--wait=live` would have the same (not very useful) semantics as before. 
</blockquote></details>

Gotcha. Like this?


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27158)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/cliflags/flags.go, line 655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LG_TZt4RfWfU8bh_5KS-r1-655:-LH4gKv77Bo-mY4cJkXt:b7u03hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfae0c2bba2248db8b2090da40b6f99409082264/pkg/cli/cliflags/flags.go#L655)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done. You didn't want to keep the original text here, right?
</blockquote></details>

Correct
___
*[pkg/cli/node.go, line 387 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27158#-LH4aY5W3hnXozhcuI8l:-LH4gH94A6bPI0f8tNxH:b2nhswy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9abd9980349cd9d9b878ac0fe49a755dc38a9bf/pkg/cli/node.go#L387)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Gotcha. Like this?
</blockquote></details>

Yes


<!-- Sent from Reviewable.io -->

tschottdorf commented:
bors r=bdarnell
